librna/rnalib.c

```diff
@@ -1067,23 +1067,60 @@ int ss_energy(rsize i, rsize j) {
   return 0;
 }
 
+static double get_mkpf_value(double energy) {
+  /* -precalculate LOOKUP_SIZE mk_pf partition function values
+     -since input values can be negative,
+      values within the range -HALF <= x < HALF
+      will be calculated (HALF = LOOKUP_SIZE / 2) */
+
+  static bool init = false;
+  static const int HALF = LOOKUP_SIZE / 2;
+  static double lookup[LOOKUP_SIZE];
+  static double divisor;
+  static int index;
+
+  if (!init) {
+    /* temperature is defined in ViennaRNA/params/basic.h
+       and can be adjusted with the -T flag at runtime */
+    divisor = GASCONST / 1000 * (temperature + K0);
+
+    // calculate mk_pf partition function values in range -HALF <= x < HALF
+    for (int i = -HALF; i < HALF; i++) {
+      lookup[i + HALF] = exp((-1.0 * i / 100.0) / divisor);
+    }
+
+    init = true;
+  }
+
+  index = energy + HALF;
+
+  if (index >= 0 && index < LOOKUP_SIZE && trunc(energy) == energy) {
+    return lookup[index];
+  } else {
+    /* if the input energy value isn't within the range
+       -HALF <= x < HALF (meaning the index isn't within the range
+       0 <= i < LOOKUP_SIZE), calculate the mk_pf value on the spot */
+    return exp((-1.0 * energy / 100.0) / divisor);
+  }
+}
 
 /*
    scales the energy value x into a partition function value
 */
-double mk_pf(double x) {
-  // temperature is defined in ViennaRNA/params/basic.h
-  return exp((-1.0 * x/100.0) / (GASCONST/1000 * (temperature + K0)));
+double mk_pf(double energy) {
```

Comment:
why does `mk_pf` only call `get_mkpf_value`? Can't we simply rename your `get_mkpf_value` function into `mk_pf` and thus replace my old implementation?

Reply:
I implemented it this way analogous to the precalculation of the `scale` function values (see PR #154), where for some reason encapsulating the code in another function and calling that function in the actual `scale` function was noticeably faster.
I could of course rename `get_mkpf_value` to `m̀k_pf` and skip the encapsulation part, but I for one assume that this weird slight speedup we saw in PR #154 when encapsulating would also happen here and also just wanted to continue that style. If you want to do all of the work in just one function we can of course do that, but in that case I would say we also do the same for `get_scale_value` and `scale`.

Reply:
oh, forgot about that curiosity. Sorry. So let's keep as is!